icu4c/source/common/putil.cpp

```diff
@@ -1175,6 +1175,18 @@ uprv_tzname(int n)
         if (ret != nullptr && uprv_strcmp(TZDEFAULT, gTimeZoneBuffer) != 0) {
             int32_t tzZoneInfoTailLen = uprv_strlen(TZZONEINFOTAIL);
             const char *tzZoneInfoTailPtr = uprv_strstr(gTimeZoneBuffer, TZZONEINFOTAIL);
+            // MacOS14 has the realpath as something like
+            // /usr/share/zoneinfo.default/Australia/Melbourne
+            // which will not have "/zoneinfo/" in the path.
+            // Therefore if we fail, we fall back to read the link which is
+            // /var/db/timezone/zoneinfo/Australia/Melbourne
+            if (tzZoneInfoTailPtr == nullptr) {
+                ssize_t size = readlink(TZDEFAULT, gTimeZoneBuffer, sizeof(gTimeZoneBuffer)-1);
```

Comment:
In other words, if the new code doesn't give us a ".../zoneinfo/..." string, then we try the old code from before PR #2323. Makes sense to me.

Alternatively, I suppose we could have looked for "zoneinfo.default" as an alternate "tail". No idea if that would be better or worse.

Apple people: Any concerns?

Reply:
I think the fallback seems reasonable.